pkg/core/resources/apis/meshservice/generate/generator.go

```diff
@@ -0,0 +1,238 @@
+		}
+		services[serviceTag] = &ms
+	}
+	return services
+}
+
+type dataplaneAndMeshService struct {
+	dataplane   model.ResourceMeta
+	meshService *meshservice_api.MeshService
+}
+
+// checkMeshServicesConsistency returns a list of dataplanes that conflict and
+// the chosen meshService
+func checkMeshServicesConsistency(
+	meshService *meshservice_api.MeshService,
+	generated []dataplaneAndMeshService,
+) ([]dataplaneAndMeshService, *meshservice_api.MeshService) {
+	if len(generated) == 0 {
+		return nil, nil
+	}
+	if meshService == nil {
+		meshService = generated[0].meshService
+	}
+	var conflicting []dataplaneAndMeshService
+	for _, generatedMeshService := range generated {
+		if servicesDiffer(meshService, generatedMeshService.meshService) {
+			conflicting = append(conflicting, generatedMeshService)
+		}
+	}
+	if len(generated) == len(conflicting) {
+		// None of the generated MeshServices match the existing, drop it and
+		// try again with the first generated. We know we won't recurse
+		// indefinitely because generated[0] won't conflict
+		return checkMeshServicesConsistency(generated[0].meshService, generated)
+	}
+	return conflicting, meshService
+}
+
+func sortDataplanes(dps []*core_mesh.DataplaneResource) []*core_mesh.DataplaneResource {
+	sorted := slices.Clone(dps)
+	slices.SortFunc(sorted, func(a, b *core_mesh.DataplaneResource) int {
+		if a, b := a.Meta.GetCreationTime(), b.Meta.GetCreationTime(); a.Before(b) {
+			return -1
+		} else if a.After(b) {
+			return 1
+		}
+		return strings.Compare(a.Meta.GetName(), b.Meta.GetName())
+	})
+	return sorted
+}
+
+func servicesDiffer(a, b *meshservice_api.MeshService) bool {
+	return !reflect.DeepEqual(a.Ports, b.Ports)
+}
+
+func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*core_mesh.DataplaneResource, meshServices []*meshservice_api.MeshServiceResource) {
+	log := g.logger.WithValues("mesh", mesh)
+	meshservicesByName := map[string][]dataplaneAndMeshService{}
+	for _, dataplane := range sortDataplanes(dataplanes) {
+		for name, ms := range g.meshServicesForDataplane(dataplane) {
+			meshservicesByName[name] = append(meshservicesByName[name], dataplaneAndMeshService{
+				dataplane:   dataplane.GetMeta(),
+				meshService: ms,
+			})
+		}
+	}
+
+	for _, meshService := range meshServices {
+		log := log.WithValues("MeshService", meshService.GetMeta().GetName())
+		conflicting, newMeshService := checkMeshServicesConsistency(meshService.Spec, meshservicesByName[meshService.GetMeta().GetName()])
+		var dps []string
+		for _, dp := range conflicting {
+			dps = append(dps, dp.dataplane.GetName())
+		}
+		if len(conflicting) > 0 {
+			log.Info("Port conflict for a kuma.io/service tag, ports must be identical across Dataplane inbounds for a given kuma.io/service", "dps", dps)
+		}
+		delete(meshservicesByName, meshService.GetMeta().GetName())
+		if newMeshService != nil && servicesDiffer(meshService.Spec, newMeshService) {
+			meshService.Spec = newMeshService
```

Comment:
I think we might need a copy of this object before update.
We are using MeshService from ReadOnlyResourceManager.
If other goroutine is reading the spec and we are writing we might get a data race.

Reply:
done